docs/docs/remark-plugin-tutorial.md

```diff
@@ -0,0 +1,283 @@
+  }
+}
+```
+
+As well, [AST Explorer](https://astexplorer.net/#/gist/d9029a2e8827265fbb9b190083b59d4d/3384f3ce6a3084e50043d0c8ce34628ed7477603) is a site that gives you a side-by-side view of the markdown and the outputted AST.
+
+## Setting up a plugin
+
+You are going to create a plugin that colors all top-level headings in the markdown with the color purple.
+
+First create a local plugin by adding a `plugins` folder in your site and generating a `package.json` file for it. As well, create an `index.js` file. In this file, we export a function that will be invoked by `gatsby-transformer-remark`:
+
+```js:title=plugins/gatsby-remark-purple-headers/index.js
+module.exports = ({ markdownAST }, pluginOptions) => {
+  // Manipulate AST
+
+  return markdownAST
+}
+```
+
+The first parameter is all of the default properties that can be used in plugins (actions, store, getNodes, schema, etc.) plus a couple just for gatsby-transformer-remark plugins. The most relevant field for our purposes is the `markdownAST` field which is destructured in the code snippet above.
+
+As with other Gatsby plugins, the 2nd parameter is the `pluginOptions` which is obtained from the definition in `gatsby-config.js` file.
+
+Finally, the function will return the `markdownAST` after the fields you wish to be edited are transformed.
+
+## Adding the plugin to your site
+
+You likely will want to grab `gatsby-source-filesystem` to inject the file nodes into Gatsby's GraphQL schema. In this example it is assumed that the Markdown files exist in a `src/data/` directory.
+
+The plugin is now initially set so you can add it as a sub-plugin inside `gatsby-transformer-remark`
+
+```js:title=gatsby-config.js
+module.exports = {
+  plugins: [
+    {
+      resolve: "gatsby-source-filesystem",
+      options: {
+        name: `data`,
+        path: `${__dirname}/src/data/`,
+      },
+    },
+    {
+      resolve: `gatsby-transformer-remark`,
+      options: {
+        plugins: [`gatsby-remark-purple-headers`],
+      },
+    },
+  ],
+}
+```
+
+If you want to add some options, you could switch to the object syntax:
+
+```js
+{
+  resolve: `gatsby-remark-purple-headers`,
+  options: {
+    // Options here
+  }
+}
+```
+
+## Find and Modify Markdown Nodes
+
+When modifying nodes, you'll want to walk the tree and then implement new functionality on specific nodes.
+
+A node module to help with is [unist-util-visit](https://github.com/syntax-tree/unist-util-visit), a walker for `unist` nodes. For reference, Unist (Unified Syntax Tree) is a standard for Markdown syntax trees and parsers that include well known parsers in the Gatsby world like Remark and MDX.
+
+As an example from `unist-util-visit`'s README file, it allows for an easy interface to visit particular nodes based on a particular type:
+
+```js
+var remark = require("remark")
+var visit = require("unist-util-visit")
+
+var tree = remark().parse("Some _emphasis_, **importance**, and `code`.")
+
+visit(tree, "text", visitor)
```

Comment:
One thing I wish was explained is the "text" parameter here. It isn't clear to me at all from looking at the `unist-util-visit` README file. Do you have an idea of what that is for? Are there other options?

Reply:
It's for listing various markdown items (heading, link, image, etc) for which a user wishes to interact with. Should this be documented further here?

Reply:
`text` is specifically blocks of text

Reply:
I think calling it out a bit more and adding info about other options would make this example even more useful!

Reply:
Done. Hopefully that helps better with understanding the `visit` function

Reply:
It's awesome!! Thanks @lannonbr 🌮